python/ray/data/preprocessors/discretizer.py

```diff
@@ -0,0 +1,360 @@
+from typing import Iterable, List, Dict, Optional, Type, Union
+
+import pandas as pd
+import numpy as np
+
+from ray.data import Dataset
+from ray.data.aggregate import Max, Min
+from ray.data.preprocessor import Preprocessor
+
+
+class _AbstractKBinsDiscretizer(Preprocessor):
+    """Abstract base class for all KBinsDiscretizers.
+
+    Essentially a thin wraper around ``pd.cut``.
+
+    Expects either ``self.stats_`` or ``self.bins`` to be set and
+    contain {column:list_of_bin_intervals}.
+    """
+
+    def _transform_pandas(self, df: pd.DataFrame):
+        def bin_values(s: pd.Series) -> pd.Series:
+            labels = self.dtypes.get(s.name) if self.dtypes else False
+            ordered = True
+            if labels:
+                if isinstance(labels, pd.CategoricalDtype):
+                    ordered = labels.ordered
+                    labels = list(labels.categories)
+                else:
+                    labels = False
+
+            bins = self.stats_ if self._is_fittable else self.bins
+            return pd.cut(
+                s,
+                bins[s.name] if isinstance(bins, dict) else bins,
+                right=self.right,
+                labels=labels,
+                ordered=ordered,
+                retbins=False,
+                include_lowest=self.include_lowest,
+                duplicates=self.duplicates,
+            )
+
+        return df.apply(bin_values, axis=0)
+
+    def _validate_bins_columns(self):
+        if isinstance(self.bins, dict) and not all(
+            col in self.bins for col in self.columns
+        ):
+            raise ValueError(
+                "If `bins` is a dictionary, all elements of `columns` must be present "
+                "in it."
+            )
+
+    def __repr__(self):
+        attr_str = ", ".join(
+            [
+                f"{attr_name}={attr_value!r}"
+                for attr_name, attr_value in vars(self).items()
+            ]
+        )
+        return f"{self.__class__.__name__}({attr_str})"
```

Comment:
Could we simplify the `repr`? The `CustomKBinsDiscretizer` `repr` looks pretty messy

> CustomKBinsDiscretizer(columns=['value_1', 'value_2'], bins={'value_1': [0, 1, 4], 'value_2': [0, 18, 35, 70]}, right=True, include_lowest=False, duplicates='raise', dtypes=None, _transform_stats='Stage 0 from_pandas_refs: 1/1 blocks executed in 0s\n* Remote wall time: 79.26ms min, 79.26ms max, 79.26ms mean, 79.26ms total\n* Remote cpu time: 79.1ms min, 79.1ms max, 79.1ms mean, 79.1ms total\n* Peak heap memory usage (MiB): 144523264000.0 min, 144523264000.0 max, 144523264000 mean\n* Output num rows: 6 min, 6 max, 6 mean, 6 total\n* Output size bytes: 224 min, 224 max, 224 mean, 224 total\n* Tasks per node: 1 min, 1 max, 1 mean; 1 nodes used\n\nStage 1 map_batches: 1/1 blocks executed in 0.09s\n* Remote wall time: 60.31ms min, 60.31ms max, 60.31ms mean, 60.31ms total\n* Remote cpu time: 60.56ms min, 60.56ms max, 60.56ms mean, 60.56ms total\n* Peak heap memory usage (MiB): 148045824000.0 min, 148045824000.0 max, 148045824000 mean\n* Output num rows: 6 min, 6 max, 6 mean, 6 total\n* Output size bytes: 224 min, 224 max, 224 mean, 224 total\n* Tasks per node: 1 min, 1 max, 1 mean; 1 nodes used\n')
`

Reply:
Ah, forgot about transform stats. Will remove it